---
layout: default
title: {{ site.name }}
---

<p><strong>scBASE</strong> is a python implementation of “soft” zero-and-one inflated beta-binomial (<strong>ZOIBB</strong>) model for estimating cellular allelic proportions in which we tackle data sparsity and variability by harnessing information derived from cell's population context.</p>

<ul>
  <li>Free software: MIT license</li>
<!--   <li>Full documentation: <a href="https://scbase.readthedocs.io/en/latest/readme.html">https://scbase.readthedocs.io/en/latest/readme.html</a></li> -->
</ul>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>First, clone the github scBASE repository under your source directory.</p>

<pre><code><strong>$</strong> git clone https://github.com/churchill-lab/scBASE.git scBASE</code></pre>

<p>Or download <code>zip</code> or <code>tar.gz</code> file from the link above and extract it in your source directory. Go to scBASE source folder. You will see the following files.</p>

<pre><code><strong>$</strong> cd scBASE
<strong>$</strong> ls
AUTHORS.rst          LICENSE              README.rst           scbase               setup.py
CONTRIBUTING.rst     MANIFEST.in          docs                 scripts              tests
HISTORY.rst          Makefile             requirements_dev.txt setup.cfg            tox.ini
</code></pre>

<p><strong>Note:</strong> We highly recommend using <a href="https://www.continuum.io/downloads">Anaconda distribution</a> of python to install all the dependencies without issues. Add the following channels if you have not already:</p>

<pre><code><strong>$</strong> conda config --add channels conda-forge
<strong>$</strong> conda config --add channels bioconda</code></pre>

<p>To avoid conflicts among dependencies, we also highly recommend using conda virtual environment:</p>

<pre><code><strong>$</strong> conda create -n scbase python=3
<strong>$</strong> source activate scbase</code></pre>

<p>Once scBASE virtual environment is created and activated, your shell prompt will show ‘(scbase)’ at the beginning to specify what virtual environment you are currently in. Now please type the following and install the dependencies for scBASE first:</p>

<pre><code>(scbase) <strong>$</strong> conda install pystan click future scipy
(scbase) <strong>$</strong> pip install loompy
</code></pre>

<p>It is now ready to install scBASE. Note compiling stan code takes a while (several minutes depending on system resource).</p>
<pre><code>(scbase) <strong>$</strong> python setup.py install</code></pre>
    
<p>That's all! Now scBASE is available. The whole installation process took 9m54s on a macbook pro with 2.7GHz Intel Core i7 and 16GB 1600MHz DDR3 RAM.</p>

<p>Once you are done using scBASE, you can go out from scBASE virtual environment anytime by deactivating it:</p>
<pre><code>(scbase) <strong>$</strong> source deactivate
</code></pre>

<h2><a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<p>We assume that users have allele-specific read counts in a <a href="http://loompy.org">loom format</a> file (where total read counts of genes are stored in the default data layer and allele-specific counts are in each allele layer using allele codes as layer keys, 'Mat' and 'Pat' for example). Or scBASE can create a loom file using allele-specific read count file(s), tab-delimited text in the following format. Users can also create count files using the results from RSEM, kallisto, or even from counting uniquely-mapping reads. The count file can contain multiple samples (or cells) as long as each sample ID is specified using a header <code>#sample_id:</code>.</p>

<pre><code>#target_id	&lt;maternal_allele_code&gt;	&lt;paternal_allele_code&gt;	total
#sample_id: &lt;Cell_ID_1&gt;
&lt;Gene_ID_1&gt;	&lt;maternal_allele_count&gt;	&lt;paternal_allele_count&gt;	&lt;total_count&gt;
&lt;Gene_ID_2&gt;	&lt;maternal_allele_count&gt;	&lt;paternal_allele_count&gt;	&lt;total_count&gt;
...
#sample_id: &lt;Cell_ID_2&gt;
&lt;Gene_ID_1&gt;	&lt;maternal_allele_count&gt;	&lt;paternal_allele_count&gt;	&lt;total_count&gt;
&lt;Gene_ID_2&gt;	&lt;maternal_allele_count&gt;	&lt;paternal_allele_count&gt;	&lt;total_count&gt;
...
</code></pre>

<p>For example, the following file was created by <a href="http://churchill-lab.github.io/emase-zero/">emase-zero</a> using reads from F1 hybrid mouse of C57BL/6J male and CAST/EiJ female (parental haplotypes denoted as <code>B</code> and <code>F</code> respectively). It has four columns: the first column is for alignment target ID's, followed by the read counts of alleles, and then, the total.</p>
<pre><code>#target_id	B	F	total
#sample_id: SRR1041728
ENSMUSG00000000001	1394.25216017	334.74783983	1729.0
ENSMUSG00000000003	0.0	0.0	0.0
ENSMUSG00000000028	843.0	2.85744524983e-63	843.0
...
#sample_id: SRR1041729
ENSMUSG00000000001	30.9238234732	0.0761765268472	31.0
ENSMUSG00000000003	0.0	0.0	0.0
ENSMUSG00000000028	554.935510723	1053.06448928	1608.0
...
</code></pre>

<h3><a id="collate" class="anchor" href="#collate" aria-hidden="true"><span class="octicon octicon-link"></span></a>subcommand: <strong>collate</strong></h3>
<p>We can collate the allele-specific count information and store it in a loom file using <code>collate</code> subcommand. scBASE can process multiple count files if the name is specified using wildcard characters (<code>*</code> or <code>?</code>).</p>
<pre><code>Usage: scbase collate [OPTIONS] &lt;indir&gt; &lt;loomfile&gt;

  Collates count or parameter files and store in a loom file
  (http://loompy.org)

Options:
  --counts                        If you are collating count files
  --params                        If you are collating count files
  --name TEXT                     File name (default: *genes*counts)
  -t, --tidfile &lt;tidfile&gt;         Name of target ID file
  -m, --model &lt;ase_model&gt; &lt;tgx_model&gt;
                                  This is a developer option used only when
                                  other models in addition to ones provided by
                                  default are available.
  -v, --verbose                   '-v' is Level 1 and '-vv' is Level 2
  --help                          Show this message and exit.
</code></pre>

<p>For example, the following command will collate allele-specific counts from all the files whose name has the pattern of <code>*.gene.counts</code> in the folder <code>/fastscratch/myproject1</code> and create a loom file named <code>proj1.loom</code>.</p>
<pre><code><strong>$</strong> scbase collate --counts -t /fastscratch/myproject1/geneID.txt --name '*.gene.counts' /fastscratch/myproject1 proj1.loom
</code></pre>
<p>We require target ID file in case the count files do not contain genes that are not expressed. For example, <code>geneID.txt</code> is a text file that contains all the Ensembl gene ID's:</p>
<pre><code>ENSMUSG00000000001
ENSMUSG00000000003
ENSMUSG00000000028
ENSMUSG00000000031
ENSMUSG00000000037
...
</code></pre>

<h3><a id="select" class="anchor" href="#select" aria-hidden="true"><span class="octicon octicon-link"></span></a>subcommand: <strong>select</strong></h3>
<p>We can select genes for fitting with ZOIBB model according to the minumum numbers of cells that satisfy the minimum number of reads expressed from a gene using <code>select</code> subcommand.</p>
<pre><code>Usage: scbase select [OPTIONS] &lt;loomfile&gt;

  Select genes if they are expressed at least &lt;min_read_count&gt; in at least
  &lt;min_cell_count&gt; cells

Options:
  --min-read-count &lt;min_read_count&gt; Min read count required
  --min-cell-count &lt;min_cell_count&gt; Min cell count required
  --layer &lt;layer_key&gt;               The layer to consider selection
  -v, --verbose                     '-v' is Level 1 and '-vv' is Level 2
  --help                            Show this message and exit.
</code></pre>
<p>For example, the following command will filter out genes that were not expressed in at least 4 reads over in at least 16 cells, and then, mark selected genes in a row attribute (<code>ra.Selected</code>) of the input loom file so the next procedure <code>run-mcmc</code> will only run on those selected genes.</p>
<pre><code><strong>$</strong> scbase select --min-read-count 4 --min-cell-count 16 proj1.loom</code></pre>

<h3><a id="mcmc" class="anchor" href="#collate" aria-hidden="true"><span class="octicon octicon-link"></span></a>subcommand: <strong>run-mcmc</strong></h3>
<p>Once input data is ready in a loom file, we can fit ZOIBB model using <code>run-mcmc</code> subcommand.</p>
<pre><code>Usage: scbase run-mcmc [OPTIONS] &lt;loomfile&gt;

  Runs MCMC (using loom file)

Options:
  --hapcode &lt;mat_hapcode&gt; &lt;pat_hapcode&gt;
                                Haplotype code for maternal and paternal
                                alleles (default: M P)
  -m, --model &lt;ase_model&gt; &lt;tgx_model&gt;
                                This is a developer option used only when
                                other models in addition to ones provided by
                                default are available.
  -s, --start &lt;gidx_start&gt;      Starting gene (row index)
  -e, --end &lt;gidx_end&gt;          Ending gene (row index)
  -o, --outfile &lt;outfile&gt;       Name of output file (default:
                                scbase.&lt;start&gt;-&lt;end&gt;.param.npz)
  -v, --verbose                 '-v' is Level 1 and '-vv' is Level 2
  --help                        Show this message and exit.
</code></pre>

<p>For example, the following command will fit the model for the first six genes in the loom file, <code>proj1.loom</code> and save the results in a file named <code>scbase.0-6.param.npz</code>. Haplotype code should be given in the order of maternal and paternal allele.</p>
<pre><code><strong>$</strong> scbase run-mcmc proj1.loom -s 0 -e 6 --hapcode F B</code></pre>

<h3><a id="submit" class="anchor" href="#submit" aria-hidden="true"><span class="octicon octicon-link"></span></a>subcommand: <strong>submit</strong></h3>
<p>Running MCMC for tens of thousands of genes would be time consuming but we can parallelize it using HPC cluster systems. We can submit jobs automatically using <code>submit</code> subcommand.</p>
<pre><code>Usage: scbase submit [OPTIONS] &lt;loomfile&gt;
  
  Submits scBASE fitting jobs to HPC clusters
  
Options:
  --hapcode &lt;mat_hapcode&gt; &lt;pat_hapcode&gt;
                                  Haplotype code for maternal and paternal
                                  alleles (default: M P)
  -m, --model &lt;ase_model&gt; &lt;tgx_model&gt;
                                  This is a developer option used only when
                                  other models in addition to ones provided by
                                  default are available.
  -c, --chunk &lt;chunk_size&gt;        Number of genes in each chunk
  -o, --outdir &lt;outdir&gt;           Folder name to store parameter files
  --systype &lt;systype&gt;             Type of HPC cluster system (default: pbs)
  --email &lt;email&gt;                 Notification E-mail
  --queue &lt;queue&gt;                 Queue name
  --mem &lt;mem&gt;                     Memory in GB (default: 16GB)
  --walltime &lt;walltime&gt;           Walltime in hours (default: 24h)
  --dryrun                        Use this when you want to rehearse your
                                  submit commands
  -v, --verbose                   '-v' is Level 1 and '-vv' is Level 2
  --help                          Show this message and exit.
</code></pre>

<p>For example, the following comand will submit jobs to PBS cluster, each fits 50 genes in <code>proj1.loom</code> file and save the results in files named <code>scbase.{gidx_start}-{gidx_end}.param.npz</code> in <code>/fastscratch/myproject1</code> folder. This will also generate temporary data files named <code>_chunk.{gidx_start}-{gidx_end}.npz</code>, which you can delete once the jobs are all done.</p>
<pre><code><strong>$</strong> scbase submit -o /fastscratch/myproject1 --systype pbs --chunk 50 --hapcode F B proj1.loom
</code></pre>

<p> Now we collate the model parameters and store in the input loom file, <code>proj1.loom</code> so that we can now have all the information in a single file.</p>
<pre><code><strong>$</strong> scbase collate --params /fastscratch/myproject1 proj1.loom
</code></pre>

<p>The <code>proj1.loom</code> file contains the following.</p>
<pre class="ipython"><code>In [1]: import loompy

In [2]: ds = loompy.connect('proj1.loom')

In [3]: ds.layers.keys()
Out[3]:
['',
 'B',
 'F',
 'lambda_k',
 'p_k',
 'pi_bk',
 'pi_mk',
 'pi_pk']

In [4]: ds.ra.keys()
Out[4]:
['GeneID',
 'Rhat_ase',
 'Rhat_tgx',
 'alpha_ase1',
 'alpha_ase2',
 'alpha_mono',
 'alpha_tgx1',
 'alpha_tgx2',
 'pi_b',
 'pi_m',
 'pi_p',
 'Selected',
 'Symbol',
 'Valid']

In [5]: ds.ca.keys()
Out[5]: ['CellID', 'CellType', 'Selected', 'Size', 'Valid']

In [6]: ds.close()

</code></pre>
<p><code>layers['lambda_k']</code> is "normalized" gene expression count per cell and <code>layers['p_k']</code> contains the final "partial pooled" allelic proportion per cell. (Note: Each layer in a loom file has a dimensionality of |num_genes|x|num_cells|.) The proportion of paternal monoallelic, bi-allelic, and maternal monoallelic cells are in <code>ra['pi_p']</code>, <code>ra['pi_b']</code>, and <code>ra['pi_m']</code>. Cell level probability of allelic expression states are available at the layers of <code>layers['pi_pk']</code>, <code>layers['pi_bk']</code>, and <code>layers['pi_mk']</code>.</p>

<div id="home">
  <h2>Announcements</h2>
  <ul class="posts">
    {% for post in site.posts %}
      <li><span>{{ post.date | date_to_string }}</span> &raquo; <a href="{{ site.github.url }}{{ post.url }}">{{ post.title }}</a></li>
    {% endfor %}
  </ul>
</div>