---
layout: default
title: {{ site.name }}
---

<p><strong>scBASE</strong> is a “soft” zero-and-one inflated beta-binomial (ZOIBB) model for cellular allelic proportions in which we tackle data sparsity and variability by harnessing information derived from cell's population context.</p>

<ul>
  <li>Free software: MIT license</li>
  <li>Full documentation: <a href="https://scbase.readthedocs.io/en/latest/readme.html">https://scbase.readthedocs.io/en/latest/readme.html</a></li>
</ul>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p><strong>Note:</strong> Although scBASE is available at <a href="https://pypi.python.org/pypi/scbase/">PyPI</a> for <code>pip install</code> or <code>easy_install</code>, we highly recommend using <a href="https://www.continuum.io/downloads">Anaconda distribution</a> of python to install all the dependencies without issues. scBASE is also available on <a href="https://anaconda.org/kbchoi/scbase">Anaconda Cloud</a>, so add the following channels if you have not already:</p>

<pre>
  <code>$ conda config --add channels conda-forge
  $ conda config --add channels bioconda</code>
</pre>

<p>To avoid conflicts among dependencies, we also highly recommend using conda virtual environment:</p>

<pre>
  <code>$ conda create -n scbase python=3
  $ source activate scbase</code>
</pre>

<p>Once scBASE virtual environment is created and activated, your shell prompt will show ‘(scbase)’ at the beginning to specify what virtual environment you are currently in. Now please type the following and install scBASE:</p>

<pre><code>(scbase) $ conda install -c kbchoi scbase
</code></pre>

<p>That's all! Now scBASE is available. Once you are done using scBASE, you can go out from scBASE virtual environment anytime:</p>

<pre><code>(scbase) $ source deactivate
</code></pre>

<h2><a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<p>We assume that users have count file(s) in the following format.</p>

<pre><code>#target_id	B	F	total
#sample_id: SRR1041728
ENSMUSG00000000001	1394.23	1394.23	2788.458371
ENSMUSG00000000003	334.771	334.771	669.541629
ENSMUSG00000000028	0	0	0.000000
...
#sample_id: SRR1041729
ENSMUSG00000000001      20.1069 20.1069 40.213792
ENSMUSG00000000003      10.8931 10.8931 21.786208
ENSMUSG00000000028      2.65648e-176    2.65648e-176    0.000000
...
</code></pre>

<p>The file has four columns in this example: the first column is for alignment target ID's, followed by the read counts of alleles, and then, the total read counts. This example is created from F1 hybrid of C57BL/6J male and CAST/EiJ female (denoted as B and F respectively). The count file can contain multiple samples (or cells) with their ID specified using '#sample_id:'. We can have multiple of these files.</p>

<h3><a id="collate" class="anchor" href="#collate" aria-hidden="true"><span class="octicon octicon-link"></span></a>collate</h3>
<p>Users can collate the count information and create a loom file using 'collate' subcommand in scBASE.</p>
<pre><code>Usage: scbase collate [OPTIONS] <indir> <loomfile>

  Utility function that collates input/output files

Options:
  --counts
  --params
  --name TEXT                     File name (wildcard characters can be used)
  -m, --model <ase_model> <tgx_model>
  -v, --verbose                   '-v' is Level 1 and '-vv' is Level 2
  --help                          Show this message and exit.
</code></pre>

<p>For example, the following command will collate counts from files named *.gene.counts under the folder /fastscratch/myproject1 and create a loom file named deng.loom.</p>
<pre><code>(scbase) $ scbase collate --counts --name *.gene.counts /fastscratch/myproject1 deng.loom
</code></pre>

<h3><a id="select" class="anchor" href="#select" aria-hidden="true"><span class="octicon octicon-link"></span></a>select</h3>
<p>Users can select genes to fit with ZOIBB model based on the minumum numbers of cells that satisfy minimum number of read count expressed from the gene using 'select' subcommand of scBASE.</p>
<pre><code>Usage: scbase select [OPTIONS] <loomfile>

  Select genes if they are expressed at least <min_read_count> in at least
  <min_cell_count> cells

Options:
  --min-read-count <min_read_count>
                                  Min read count required
  --min-cell-count <min_cell_count>
                                  Min cell count required
  --layer <layer_key>             The layer to consider selection
  -v, --verbose                   the more times listed, the more output
  --help                          Show this message and exit.
</code></pre>
<p>For example, the following command will mark selected genes as a row attribute of the input loom file (ra.Selected) and the fitting process in the next procedure will only run on those selected genes.</p>
<pre><code>(scbase) $ scbase select --min-read-count 4 --min-cell-count 16 deng.loom</code></pre>

<h3><a id="mcmc" class="anchor" href="#collate" aria-hidden="true"><span class="octicon octicon-link"></span></a>run_mcmc</h3>
<p>Once input data is ready in a loom file format, we can fit ZOIBB model using 'run_mcmc' subcommand in scBASE.</p>
<pre><code>Usage: scbase run_mcmc [OPTIONS] <npzfile>

  Run MCMC (using npz file)

Options:
  --hapcode <mat_hapcode> <pat_hapcode>
  -m, --model <ase_model> <tgx_model>
  -s, --start <gidx_start>         Starting gene (row index)
  -e, --end <gidx_end>             Ending gene (row index)
  -o, --outfile <outfile>
  -v, --verbose                   '-v' is Level 1 and '-vv' is Level 2
  --help                          Show this message and exit.
</code></pre>

<p>For example, the following command will fit the model for the first six genes in the loom file, deng.loom. Haplotype information is given in the order of maternal and paternal allele.</p>
<pre><code>(scbase) $ scbase run_mcmc deng.loom -s 0 -e 6 --hapcode F B -v</code></pre>

<h3><a id="submit" class="anchor" href="#submit" aria-hidden="true"><span class="octicon octicon-link"></span></a>submit</h3>
<pre><code>Usage: scbase submit [OPTIONS] <loomfile>
  
    Submit jobs to HPC clusters
  
  Options:
    --hapcode <mat_hapcode> <pat_hapcode>
    -m, --model <ase_model> <tgx_model>
    -c, --chunk <chunk_size>        Chunk size
    -o, --outdir <outdir>
    --systype <systype>             Type of HPC cluster system (default: pbs)
    --email <email>                 Notification E-mail
    --queue <queue>                 Queue name
    --mem <mem>                     Memory in GB (default: 16GB)
    --walltime <walltime>           Walltime in hours (default: 24h
    --dryrun
    -v, --verbose                   '-v' is Level 1 and '-vv' is Level 2
    --help                          Show this message and exit.
  </code></pre>

<p>For example, the following comand will submit jobs to PBS cluster, each fits 50 genes in deng.loom file. </p>
<pre><code>(scbase) $ scbase submit -o /fastscratch/myproject1 --systype pbs --chunk 50 --hapcode F B deng.loom
</code></pre>

<p>The ZOIBB model parameters will be stored using a file name, scbase.{gidx_start}-{gidx_end}.param.npz, at the specified output folder, e.g., /fastscratch/myproject1. We will collate the results and store in the input file, deng.loom.</p>
<pre><code>(scbase) $ scbase collate --params /fastscratch/myproject1 deng.loom
</code></pre>

<div id="home">
  <h2>Announcements</h2>
  <ul class="posts">
    {% for post in site.posts %}
      <li><span>{{ post.date | date_to_string }}</span> &raquo; <a href="{{ site.github.url }}{{ post.url }}">{{ post.title }}</a></li>
    {% endfor %}
  </ul>
</div>